controllers/gitrepository_controller.go

```diff
@@ -273,6 +279,9 @@ func (r *GitRepositoryReconciler) reconcile(ctx context.Context, obj *sourcev1.G
 func (r *GitRepositoryReconciler) reconcileStorage(ctx context.Context,
 	obj *sourcev1.GitRepository, _ *git.Commit, _ *artifactSet, _ string) (sreconcile.Result, error) {
 	// Garbage collect previous advertised artifact(s) from storage
+	// Abort if it takes more than 1/3 the requeue interval to avoid stalling the worker.
+	ctx, cancel := context.WithTimeout(ctx, time.Duration(obj.Spec.Interval.Duration.Seconds()/3))
+	defer cancel()
```

Comment:
Maybe this should be along with the rest of the GC code in `garbageCollect()`.
Not sure if it makes any difference to keep it here.

Reply:
Regarding the timeout, 1/3 of the interval may not be appropriate. If my interval is set to 3 hours, this timeout would be 1 hour. Allowing GC to run for 1 hour may not do any good. Maybe a fixed short value would be more appropriate. If someone triggers a reconciliation, and something goes wrong, in the aforementioned scenario, they have to wait for an hour for the GC to timeout and find out that something went wrong. A short predictable timeout would make things more responsive and deterministic.

Reply:
How about a predefined duration such as 20s and comparing that with 1/3 of the interval, and choosing whichever is shorter?

Reply:
Not sure why it needs to be related to the interval. If a reconciliation is running, it should complete regardless of the interval. If something takes 2 minutes to run but their interval is 1 minute, the reconciliation need not compromise based on the interval value. Maybe it should be based on expected responsiveness of the system, which may vary from user to user. But maybe we can think of something that would be appropriate enough since the scenario of too many things to delete and timing out won't be usual. As we add new artifact, we delete other artifacts. Most of the time, a short time would be enough. Only in strange cases where too many files were injected in the filesystem and deleting that may take some time, there would be cases of timeout. But even in those scenarios, a short enough timeout may be able to cleanup everything eventually.